File: src/commands/prompts.ts
/**
 * Custom Prompts System - User-definable slash commands from Markdown files.
 *
 * Inspired by OpenAI Codex's prompts system. Prompts are discovered from:
 * - ~/.composer/prompts/*.md (user prompts)
 * - .composer/prompts/*.md (project prompts)
 *
 * Additionally, Composer treats Markdown files in `.composer/commands/*.md` and
 * `~/.composer/commands/*.md` as prompts. This lets teams colocate "prompt-like"
 * commands alongside other command templates.
 *
 * Each prompt is a Markdown file with optional YAML frontmatter:
 *
 * ```markdown
 * ---
 * description: Request a concise git diff review
 * argument-hint: FILE=<path> [FOCUS=<section>]
 * ---
 *
 * Review the code in $FILE. Pay special attention to $FOCUS.
 * ```
 *
 * Placeholders:
 * - Positional: $1, $2, ..., $9 (from space-separated args)
 * - $ARGUMENTS: All positional arguments joined by space
 * - Named: $FILE, $TICKET_ID (from KEY=value pairs)
 * - Escape: $$ produces a literal $
 */

import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { basename, join } from "node:path";
import { createLogger } from "../utils/logger.js";

const logger = createLogger("commands:prompts");

function resolveHomeDirectory(): string {
	// Prefer env vars so tests (and some shells) can override HOME reliably.
	return process.env.HOME || process.env.USERPROFILE || homedir();
}

/**
 * Prompt definition from YAML frontmatter.
 */
export interface PromptDefinition {
	/** Unique prompt name (derived from filename) */
	name: string;
	/** Short description shown in slash popup */
	description?: string;
	/** Alternative names for this prompt */
	aliases?: string[];
	/** Hint for expected arguments */
	argumentHint?: string;
	/** Full markdown body (content after frontmatter) */
	body: string;
	/** Source file path */
	sourcePath: string;
	/** Source type: 'user' or 'project' */
	sourceType: "user" | "project";
	/** Named placeholders found in the body */
	namedPlaceholders: string[];
	/** Whether body uses positional placeholders */
	hasPositionalPlaceholders: boolean;
}

/**
 * Result of parsing prompt arguments.
 */
export interface ParsedPromptArgs {
	/** Positional arguments ($1, $2, etc.) */
	positional: string[];
	/** Named arguments (KEY=value) */
	named: Record<string, string>;
}

/**
 * Parse YAML frontmatter from markdown content.
 */
function parseFrontmatter(content: string): {
	frontmatter: Record<string, unknown>;
	body: string;
} {
	const frontmatterRegex = /^---\s*\n([\s\S]*?)\n---\s*\n?([\s\S]*)$/;
	const match = content.match(frontmatterRegex);

	if (!match) {
		return { frontmatter: {}, body: content };
	}

	const [, yamlContent, body] = match;
	const frontmatter: Record<string, unknown> = {};

	// Simple YAML parser for common patterns (key: value and arrays)
	const lines = yamlContent.split("\n");
	let currentKey: string | null = null;
	let currentArray: string[] | null = null;

	for (const line of lines) {
		const trimmed = line.trim();
		if (!trimmed || trimmed.startsWith("#")) continue;

		// Array item
		if (trimmed.startsWith("- ") && currentKey && currentArray) {
			currentArray.push(trimmed.slice(2).trim());
			continue;
		}

		// Save previous array
		if (currentKey && currentArray) {
			frontmatter[currentKey] = currentArray;
			currentKey = null;
			currentArray = null;
		}

		const colonIndex = trimmed.indexOf(":");
		if (colonIndex <= 0) continue;
		const key = trimmed.slice(0, colonIndex).trim();
		let value = trimmed.slice(colonIndex + 1).trim();

		if (value === "") {
			currentKey = key;
			currentArray = [];
			continue;
		}

		// Remove surrounding quotes if present
		if (
			(value.startsWith('"') && value.endsWith('"')) ||
			(value.startsWith("'") && value.endsWith("'"))
		) {
			value = value.slice(1, -1);
		}

		frontmatter[key] = value;
	}

	// Save trailing array
	if (currentKey && currentArray) {
		frontmatter[currentKey] = currentArray;
	}

	return { frontmatter, body };
}

function isValidPromptName(value: string): boolean {
	return /^[a-zA-Z][a-zA-Z0-9_-]*$/.test(value);
}

function parseAliases(raw: unknown): string[] | undefined {
	const values: string[] = [];

	if (Array.isArray(raw)) {
		for (const item of raw) {
			if (typeof item === "string") values.push(item);
		}
	} else if (typeof raw === "string") {
		// Accept simple comma-separated strings as well.
		values.push(
			...raw
				.split(",")
				.map((part) => part.trim())
				.filter(Boolean),
		);
	}

	const normalized = new Set<string>();
	const cleaned: string[] = [];
	for (const value of values) {
		if (!isValidPromptName(value)) continue;
		const key = value.toLowerCase();
		if (normalized.has(key)) continue;
		normalized.add(key);
		cleaned.push(value);
	}

	return cleaned.length > 0 ? cleaned : undefined;
}

/**
 * Extract named placeholders from prompt body.
 * Named placeholders are $UPPERCASE_IDENTIFIERS (not $1-$9 or $$).
 */
function extractNamedPlaceholders(body: string): string[] {
	const pattern = /\$([A-Z][A-Z0-9_]*)/g;
	const placeholders = new Set<string>();

	for (const match of body.matchAll(pattern)) {
		const name = match[1];
		// Skip special placeholders
		if (name === "ARGUMENTS") continue;
		placeholders.add(name);
	}

	return Array.from(placeholders).sort();
}

/**
 * Check if body uses positional placeholders ($1-$9 or $ARGUMENTS).
 */
function hasPositionalPlaceholders(body: string): boolean {
	return /\$[1-9]|\$ARGUMENTS/.test(body);
}

/**
 * Load a single prompt from a markdown file.
 */
function loadPromptFromFile(
	filePath: string,
	sourceType: "user" | "project",
): PromptDefinition | null {
	try {
		const content = readFileSync(filePath, "utf-8");
		const { frontmatter, body } = parseFrontmatter(content);

		// Derive name from filename (without .md extension), but allow override.
		const fileName = basename(filePath, ".md");
		const frontmatterName =
			typeof frontmatter.name === "string" ? frontmatter.name.trim() : "";
		const nameCandidate = frontmatterName || fileName;
		const name = isValidPromptName(nameCandidate) ? nameCandidate : fileName;

		if (!name || name.startsWith(".")) {
			logger.debug("Skipping hidden or invalid prompt file", { filePath });
			return null;
		}

		const aliases = parseAliases(frontmatter.aliases);
		const filteredAliases = aliases?.filter(
			(alias) => alias.toLowerCase() !== name.toLowerCase(),
		);

		const prompt: PromptDefinition = {
			name,
			description:
				typeof frontmatter.description === "string"
					? frontmatter.description
					: undefined,
			aliases: filteredAliases?.length ? filteredAliases : undefined,
			argumentHint:
				typeof frontmatter["argument-hint"] === "string"
					? frontmatter["argument-hint"]
					: typeof frontmatter.argument_hint === "string"
						? frontmatter.argument_hint
						: undefined,
			body: body.trim(),
			sourcePath: filePath,
			sourceType,
			namedPlaceholders: extractNamedPlaceholders(body),
			hasPositionalPlaceholders: hasPositionalPlaceholders(body),
		};

		logger.debug("Loaded prompt", {
			name: prompt.name,
			sourceType,
			namedPlaceholders: prompt.namedPlaceholders,
		});

		return prompt;
	} catch (err) {
		logger.warn("Error loading prompt", {
			filePath,
			error: err instanceof Error ? err.message : String(err),
		});
		return null;
	}
}

/**
 * Scan a directory for prompt markdown files.
 */
function scanPromptsDirectory(
	dir: string,
	sourceType: "user" | "project",
): PromptDefinition[] {
	if (!existsSync(dir)) {
		return [];
	}

	const prompts: PromptDefinition[] = [];

	try {
		const entries = readdirSync(dir);
		for (const entry of entries) {
			if (!entry.endsWith(".md")) continue;

			const entryPath = join(dir, entry);
			try {
				const stat = statSync(entryPath);
				if (!stat.isFile() && !stat.isSymbolicLink()) continue;

				const prompt = loadPromptFromFile(entryPath, sourceType);
				if (prompt) {
					prompts.push(prompt);
				}
			} catch {
				// Skip inaccessible files
			}
		}
	} catch (err) {
		logger.warn("Error scanning prompts directory", {
			dir,
			error: err instanceof Error ? err.message : String(err),
		});
	}

	return prompts;
}

/**
 * Load all available prompts from user and project directories.
 *
 * @param workspaceDir - The current workspace/project directory
 * @returns Array of loaded prompts (project prompts override user prompts by name)
 */
export function loadPrompts(workspaceDir: string): PromptDefinition[] {
	const homeDir = resolveHomeDirectory();
	const userPromptsDir = join(homeDir, ".composer", "prompts");
	const userCommandsDir = join(homeDir, ".composer", "commands");
	const projectPromptsDir = join(workspaceDir, ".composer", "prompts");
	const projectCommandsDir = join(workspaceDir, ".composer", "commands");

	logger.debug("Scanning for prompts", {
		userPromptsDir,
		projectPromptsDir,
		userCommandsDir,
		projectCommandsDir,
	});

	// Commands dir first, prompts dir second (prompts override commands by name).
	const userPrompts = [
		...scanPromptsDirectory(userCommandsDir, "user"),
		...scanPromptsDirectory(userPromptsDir, "user"),
	];
	const projectPrompts = [
		...scanPromptsDirectory(projectCommandsDir, "project"),
		...scanPromptsDirectory(projectPromptsDir, "project"),
	];

	// Project prompts override user prompts by name
	const promptMap = new Map<string, PromptDefinition>();

	for (const prompt of userPrompts) {
		promptMap.set(prompt.name.toLowerCase(), prompt);
	}

	for (const prompt of projectPrompts) {
		const existing = promptMap.get(prompt.name.toLowerCase());
		if (existing) {
			logger.debug("Project prompt overrides user prompt", {
				name: prompt.name,
			});
		}
		promptMap.set(prompt.name.toLowerCase(), prompt);
	}

	const allPrompts = Array.from(promptMap.values());

	// Sort by name for consistent ordering
	allPrompts.sort((a, b) => a.name.localeCompare(b.name));

	logger.info("Finished loading prompts", {
		total: allPrompts.length,
		user: userPrompts.length,
		project: projectPrompts.length,
	});

	return allPrompts;
}

/**
 * Find a prompt by name (case-insensitive).
 */
export function findPrompt(
	prompts: PromptDefinition[],
	name: string,
): PromptDefinition | undefined {
	const normalizedName = name.toLowerCase();
	return prompts.find((p) => {
		if (p.name.toLowerCase() === normalizedName) return true;
		return p.aliases?.some((alias) => alias.toLowerCase() === normalizedName);
	});
}

/**
 * Parse arguments from a command invocation.
 *
 * Supports:
 * - Positional args: separated by whitespace
 * - Named args: KEY=value or KEY="value with spaces"
 *
 * @param argString - The argument string after the command name
 * @returns Parsed positional and named arguments
 */
export function parsePromptArgs(argString: string): ParsedPromptArgs {
	const positional: string[] = [];
	const named: Record<string, string> = {};

	if (!argString.trim()) {
		return { positional, named };
	}

	// Tokenize respecting quoted strings
	const tokens: string[] = [];
	let current = "";
	let inQuotes = false;
	let quoteChar = "";

	for (let i = 0; i < argString.length; i++) {
		const char = argString[i];

		if (!inQuotes && (char === '"' || char === "'")) {
			inQuotes = true;
			quoteChar = char;
			continue;
		}

		if (inQuotes && char === quoteChar) {
			inQuotes = false;
			quoteChar = "";
			continue;
		}

		if (!inQuotes && /\s/.test(char)) {
			if (current) {
				tokens.push(current);
				current = "";
			}
			continue;
		}

		current += char;
	}

	if (current) {
		tokens.push(current);
	}

	// Separate named and positional arguments
	for (const token of tokens) {
		const eqIndex = token.indexOf("=");
		if (eqIndex > 0 && /^[A-Z][A-Z0-9_]*$/.test(token.slice(0, eqIndex))) {
			// Named argument (KEY=value)
			const key = token.slice(0, eqIndex);
			const value = token.slice(eqIndex + 1);
			named[key] = value;
		} else {
			// Positional argument
			positional.push(token);
		}
	}

	return { positional, named };
}

/**
 * Validate that all required named placeholders are provided.
 *
 * @returns Error message if validation fails, null if valid
 */
export function validatePromptArgs(
	prompt: PromptDefinition,
	args: ParsedPromptArgs,
): string | null {
	const missing: string[] = [];

	for (const placeholder of prompt.namedPlaceholders) {
		if (!(placeholder in args.named)) {
			missing.push(placeholder);
		}
	}

	if (missing.length > 0) {
		return `Missing required arguments: ${missing.join(", ")}`;
	}

	return null;
}

/**
 * Render a prompt with the given arguments.
 *
 * Substitutes:
 * - $1, $2, ..., $9 with positional arguments
 * - $ARGUMENTS with all positional arguments joined by space
 * - $NAME with named argument values
 * - $$ with a literal $
 */
export function renderPrompt(
	prompt: PromptDefinition,
	args: ParsedPromptArgs,
): string {
	let result = prompt.body;

	// First, escape $$ to a placeholder
	const escapeMarker = "\x00DOLLAR\x00";
	result = result.replace(/\$\$/g, escapeMarker);

	// Substitute $ARGUMENTS
	result = result.replace(/\$ARGUMENTS/g, args.positional.join(" "));

	// Substitute positional arguments $1-$9
	for (let i = 1; i <= 9; i++) {
		const pattern = new RegExp(`\\$${i}`, "g");
		const value = args.positional[i - 1] ?? "";
		result = result.replace(pattern, value);
	}

	// Substitute named arguments
	for (const [key, value] of Object.entries(args.named)) {
		const pattern = new RegExp(`\\$${key}`, "g");
		result = result.replace(pattern, value);
	}

	// Restore escaped dollars
	result = result.replace(new RegExp(escapeMarker, "g"), "$");

	return result;
}

/**
 * Format prompt for display in a list.
 */
export function formatPromptListItem(prompt: PromptDefinition): string {
	const source = prompt.sourceType === "user" ? "(user)" : "(project)";
	const desc = prompt.description ?? "(no description)";
	const aliases = prompt.aliases?.length
		? ` (aliases: ${prompt.aliases.join(", ")})`
		: "";
	return `${prompt.name}${aliases} ${source} - ${desc}`;
}

/**
 * Get usage hint for a prompt.
 */
export function getPromptUsageHint(prompt: PromptDefinition): string {
	const parts: string[] = [`/prompts:${prompt.name}`];

	if (prompt.argumentHint) {
		parts.push(prompt.argumentHint);
	} else if (prompt.namedPlaceholders.length > 0) {
		parts.push(prompt.namedPlaceholders.map((p) => `${p}=<value>`).join(" "));
	} else if (prompt.hasPositionalPlaceholders) {
		parts.push("<args...>");
	}

	return parts.join(" ");
}

File: packages/tui-rs/src/prompts.rs
//! Custom Prompts System - User-definable slash commands from Markdown files.
//!
//! Prompts are discovered from:
//! - ~/.composer/prompts/*.md (user prompts)
//! - .composer/prompts/*.md (project prompts)
//!
//! Each prompt is a Markdown file with optional YAML frontmatter:
//!
//! ```markdown
//! ---
//! description: Request a concise git diff review
//! argument-hint: FILE=<path> [FOCUS=<section>]
//! ---
//!
//! Review the code in $FILE. Pay special attention to $FOCUS.
//! ```
//!
//! Placeholders:
//! - Positional: $1, $2, ..., $9 (from space-separated args)
//! - $ARGUMENTS: All positional arguments joined by space
//! - Named: $FILE, $TICKET_ID (from KEY=value pairs)
//! - Escape: $$ produces a literal $

use crate::path_utils::{legacy_composer_home_dir, maestro_home_dir};
use regex::Regex;
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::LazyLock;

/// A prompt definition loaded from a markdown file.
#[derive(Debug, Clone)]
pub struct PromptDefinition {
    /// Unique prompt name (derived from filename)
    pub name: String,
    /// Short description shown in slash popup
    pub description: Option<String>,
    /// Hint for expected arguments
    pub argument_hint: Option<String>,
    /// Full markdown body (content after frontmatter)
    pub body: String,
    /// Source file path
    pub source_path: PathBuf,
    /// Source type: "user" or "project"
    pub source_type: PromptSource,
    /// Named placeholders found in the body
    pub named_placeholders: Vec<String>,
    /// Whether body uses positional placeholders
    pub has_positional_placeholders: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PromptSource {
    User,
    Plugin,
    Project,
}

impl PromptSource {
    #[must_use]
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::User => "user",
            Self::Plugin => "plugin",
            Self::Project => "project",
        }
    }
}

/// Parsed arguments from a prompt invocation.
#[derive(Debug, Clone, Default)]
pub struct ParsedArgs {
    /// Positional arguments ($1, $2, etc.)
    pub positional: Vec<String>,
    /// Named arguments (KEY=value)
    pub named: HashMap<String, String>,
}

static NAMED_PLACEHOLDER_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\$([A-Z][A-Z0-9_]*)").unwrap());
static POSITIONAL_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\$[1-9]|\$ARGUMENTS").unwrap());
/// Regex for parsing named arguments (KEY=value) from argument strings
static NAMED_ARG_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^([A-Z][A-Z0-9_]*)=(.*)$").unwrap());

/// Parse YAML frontmatter from markdown content.
fn parse_frontmatter(content: &str) -> (HashMap<String, String>, &str) {
    let mut frontmatter = HashMap::new();

    if !content.starts_with("---") {
        return (frontmatter, content);
    }

    let lines: Vec<&str> = content.lines().collect();
    if lines.len() < 2 {
        return (frontmatter, content);
    }

    let mut end_idx = None;
    for (i, line) in lines.iter().enumerate().skip(1) {
        if line.trim() == "---" {
            end_idx = Some(i);
            break;
        }
    }

    let Some(end_idx) = end_idx else {
        return (frontmatter, content);
    };

    // Parse simple key: value pairs
    for line in &lines[1..end_idx] {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        if let Some(colon_idx) = line.find(':') {
            let key = line[..colon_idx].trim().to_string();
            let mut value = line[colon_idx + 1..].trim().to_string();
            // Remove surrounding quotes
            if (value.starts_with('"') && value.ends_with('"'))
                || (value.starts_with('\'') && value.ends_with('\''))
            {
                value = value[1..value.len() - 1].to_string();
            }
            frontmatter.insert(key, value);
        }
    }

    // Body is everything after the closing ---
    let body_start = lines[..=end_idx].iter().map(|l| l.len() + 1).sum::<usize>();
    let body = if body_start < content.len() {
        &content[body_start..]
    } else {
        ""
    };

    (frontmatter, body)
}

/// Extract named placeholders from prompt body.
fn extract_named_placeholders(body: &str) -> Vec<String> {
    let mut placeholders: Vec<String> = NAMED_PLACEHOLDER_RE
        .captures_iter(body)
        .filter_map(|cap| {
            let name = cap.get(1)?.as_str();
            if name == "ARGUMENTS" {
                None
            } else {
                Some(name.to_string())
            }
        })
        .collect();
    placeholders.sort();
    placeholders.dedup();
    placeholders
}

/// Check if body uses positional placeholders.
fn has_positional_placeholders(body: &str) -> bool {
    POSITIONAL_RE.is_match(body)
}

/// Load a single prompt from a markdown file.
fn load_prompt_from_file(path: &Path, source_type: PromptSource) -> Option<PromptDefinition> {
    let content = fs::read_to_string(path).ok()?;
    let name = path.file_stem()?.to_string_lossy().to_string();

    if name.starts_with('.') {
        return None;
    }

    let (frontmatter, body) = parse_frontmatter(&content);
    let body = body.trim().to_string();

    Some(PromptDefinition {
        name,
        description: frontmatter.get("description").cloned(),
        argument_hint: frontmatter
            .get("argument-hint")
            .or_else(|| frontmatter.get("argument_hint"))
            .cloned(),
        named_placeholders: extract_named_placeholders(&body),
        has_positional_placeholders: has_positional_placeholders(&body),
        body,
        source_path: path.to_path_buf(),
        source_type,
    })
}

/// Scan a directory for prompt markdown files.
fn scan_prompts_directory(dir: &Path, source_type: PromptSource) -> Vec<PromptDefinition> {
    let mut prompts = Vec::new();

    if !dir.exists() {
        return prompts;
    }

    if let Ok(entries) = fs::read_dir(dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.extension().is_some_and(|e| e == "md") {
                if let Some(prompt) = load_prompt_from_file(&path, source_type) {
                    prompts.push(prompt);
                }
            }
        }
    }

    prompts
}

/// Load all available prompts from user and project directories.
///
/// Later paths override earlier ones by name (project maestro wins).
#[must_use]
pub fn load_prompts(workspace_dir: &Path) -> Vec<PromptDefinition> {
    load_prompts_with_plugin_dirs(workspace_dir, &[])
}

/// Load prompts while including trusted plugin command directories.
#[must_use]
pub fn load_prompts_with_plugin_dirs(
    workspace_dir: &Path,
    plugin_dirs: &[PathBuf],
) -> Vec<PromptDefinition> {
    let mut dirs: Vec<(PathBuf, PromptSource)> = Vec::new();

    if let Some(home) = legacy_composer_home_dir() {
        dirs.push((home.join("commands"), PromptSource::User));
        dirs.push((home.join("prompts"), PromptSource::User));
    }
    if let Some(home) = maestro_home_dir() {
        dirs.push((home.join("commands"), PromptSource::User));
        dirs.push((home.join("prompts"), PromptSource::User));
    }
    dirs.extend(
        plugin_dirs
            .iter()
            .cloned()
            .map(|path| (path, PromptSource::Plugin)),
    );

    dirs.push((
        workspace_dir.join(".composer").join("commands"),
        PromptSource::Project,
    ));
    dirs.push((
        workspace_dir.join(".composer").join("prompts"),
        PromptSource::Project,
    ));
    dirs.push((
        workspace_dir.join(".agents").join("commands"),
        PromptSource::Project,
    ));
    dirs.push((
        workspace_dir.join(".maestro").join("commands"),
        PromptSource::Project,
    ));
    dirs.push((
        workspace_dir.join(".maestro").join("prompts"),
        PromptSource::Project,
    ));

    let mut prompt_map: HashMap<String, PromptDefinition> = HashMap::new();
    for (dir, source) in dirs {
        for prompt in scan_prompts_directory(&dir, source) {
            prompt_map.insert(prompt.name.to_lowercase(), prompt);
        }
    }

    let mut prompts: Vec<_> = prompt_map.into_values().collect();
    prompts.sort_by(|a, b| a.name.cmp(&b.name));
    prompts
}

/// Format a prompt/command template for slash invocation.
pub fn format_prompt_invoke(prompt: &PromptDefinition, raw_args: &str) -> Result<String, String> {
    let args = parse_args(raw_args);
    if prompt.has_positional_placeholders || !prompt.named_placeholders.is_empty() {
        validate_args(prompt, &args)?;
        return Ok(render_prompt(prompt, &args));
    }
    let body = prompt.body.trim();
    let args_trim = raw_args.trim();
    if args_trim.is_empty() {
        Ok(body.to_string())
    } else if body.is_empty() {
        Ok(args_trim.to_string())
    } else {
        Ok(format!("{body}\n\n---\n\n{args_trim}"))
    }
}

/// Find a prompt by name (case-insensitive).
#[must_use]
pub fn find_prompt<'a>(
    prompts: &'a [PromptDefinition],
    name: &str,
) -> Option<&'a PromptDefinition> {
    let name_lower = name.to_lowercase();
    prompts.iter().find(|p| p.name.to_lowercase() == name_lower)
}

/// Parse arguments from a command invocation.
pub fn parse_args(arg_string: &str) -> ParsedArgs {
    let mut result = ParsedArgs::default();

    if arg_string.trim().is_empty() {
        return result;
    }

    // Tokenize respecting quoted strings
    let mut tokens = Vec::new();
    let mut current = String::new();
    let mut in_quotes = false;
    let mut quote_char = ' ';

    for ch in arg_string.chars() {
        if !in_quotes && (ch == '"' || ch == '\'') {
            in_quotes = true;
            quote_char = ch;
            continue;
        }

        if in_quotes && ch == quote_char {
            in_quotes = false;
            continue;
        }

        if !in_quotes && ch.is_whitespace() {
            if !current.is_empty() {
                tokens.push(std::mem::take(&mut current));
            }
            continue;
        }

        current.push(ch);
    }

    if !current.is_empty() {
        tokens.push(current);
    }

    // Separate named and positional arguments
    for token in tokens {
        if let Some(caps) = NAMED_ARG_RE.captures(&token) {
            let key = caps.get(1).unwrap().as_str().to_string();
            let value = caps.get(2).unwrap().as_str().to_string();
            result.named.insert(key, value);
        } else {
            result.positional.push(token);
        }
    }

    result
}

/// Validate that all required named placeholders are provided.
pub fn validate_args(prompt: &PromptDefinition, args: &ParsedArgs) -> Result<(), String> {
    let missing: Vec<_> = prompt
        .named_placeholders
        .iter()
        .filter(|p| !args.named.contains_key(*p))
        .collect();

    if missing.is_empty() {
        Ok(())
    } else {
        Err(format!(
            "Missing required arguments: {}",
            missing
                .iter()
                .map(|s| s.as_str())
                .collect::<Vec<_>>()
                .join(", ")
        ))
    }
}

/// Render a prompt with the given arguments.
#[must_use]
pub fn render_prompt(prompt: &PromptDefinition, args: &ParsedArgs) -> String {
    let mut result = prompt.body.clone();

    // Escape $$ to a placeholder
    let escape_marker = "\x00DOLLAR\x00";
    result = result.replace("$$", escape_marker);

    // Substitute $ARGUMENTS
    result = result.replace("$ARGUMENTS", &args.positional.join(" "));

    // Substitute positional arguments $1-$9
    for i in 1..=9 {
        let pattern = format!("${i}");
        let value = args
            .positional
            .get(i - 1)
            .map_or("", std::string::String::as_str);
        result = result.replace(&pattern, value);
    }

    // Substitute named arguments
    for (key, value) in &args.named {
        let pattern = format!("${key}");
        result = result.replace(&pattern, value);
    }

    // Restore escaped dollars
    result = result.replace(escape_marker, "$");

    result
}

/// Format prompt for display in a list.
#[must_use]
pub fn format_prompt_list_item(prompt: &PromptDefinition) -> String {
    let source = match prompt.source_type {
        PromptSource::User => "(user)",
        PromptSource::Plugin => "(plugin)",
        PromptSource::Project => "(project)",
    };
    let desc = prompt.description.as_deref().unwrap_or("(no description)");
    format!("{} {} - {}", prompt.name, source, desc)
}

/// Get usage hint for a prompt.
#[must_use]
pub fn get_usage_hint(prompt: &PromptDefinition) -> String {
    let mut parts = vec![format!("/{}", prompt.name)];

    if let Some(hint) = &prompt.argument_hint {
        parts.push(hint.clone());
    } else if !prompt.named_placeholders.is_empty() {
        let args: Vec<_> = prompt
            .named_placeholders
            .iter()
            .map(|p| format!("{p}=<value>"))
            .collect();
        parts.push(args.join(" "));
    } else if prompt.has_positional_placeholders {
        parts.push("<args...>".to_string());
    }

    parts.join(" ")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_load_prompts_prefers_maestro_over_composer() {
        let temp = tempfile::TempDir::new().unwrap();
        let workspace = temp.path();
        let composer = workspace.join(".composer").join("prompts");
        let maestro = workspace.join(".maestro").join("prompts");
        let commands = workspace.join(".maestro").join("commands");
        fs::create_dir_all(&composer).unwrap();
        fs::create_dir_all(&maestro).unwrap();
        fs::create_dir_all(&commands).unwrap();
        fs::write(
            composer.join("review.md"),
            "---
description: composer
---
composer body
",
        )
        .unwrap();
        fs::write(
            maestro.join("review.md"),
            "---
description: maestro
---
maestro body
",
        )
        .unwrap();
        fs::write(
            commands.join("ship.md"),
            "---
description: ship
---
Ship $ARGUMENTS
",
        )
        .unwrap();
        let prompts = load_prompts(workspace);
        assert_eq!(
            find_prompt(&prompts, "review").unwrap().body.trim(),
            "maestro body"
        );
        assert!(find_prompt(&prompts, "ship").is_some());
    }

    #[test]
    fn loads_markdown_commands_from_plugins() {
        let temp = tempfile::tempdir().unwrap();
        let workspace = temp.path().join("workspace");
        let commands = temp.path().join("plugin").join("commands");
        fs::create_dir_all(&commands).unwrap();
        fs::write(
            commands.join("plugin-review.md"),
            "---\ndescription: Review from plugin\n---\nReview $ARGUMENTS",
        )
        .unwrap();

        let prompts = load_prompts_with_plugin_dirs(&workspace, &[commands]);
        let prompt = find_prompt(&prompts, "plugin-review").expect("plugin prompt");

        assert_eq!(prompt.source_type, PromptSource::Plugin);
        assert_eq!(prompt.description.as_deref(), Some("Review from plugin"));
    }

    #[test]
    fn test_format_prompt_invoke_appends_args_without_placeholders() {
        let prompt = PromptDefinition {
            name: "commit".to_string(),
            description: None,
            argument_hint: None,
            body: "Create a conventional commit.".to_string(),
            source_path: PathBuf::new(),
            source_type: PromptSource::User,
            named_placeholders: vec![],
            has_positional_placeholders: false,
        };
        let out = format_prompt_invoke(&prompt, "fix typo").unwrap();
        assert!(out.contains("Create a conventional commit."));
        assert!(out.contains("fix typo"));
    }

    #[test]
    fn test_parse_args_positional() {
        let args = parse_args("foo bar baz");
        assert_eq!(args.positional, vec!["foo", "bar", "baz"]);
        assert!(args.named.is_empty());
    }

    #[test]
    fn test_parse_args_named() {
        let args = parse_args("FILE=main.ts FOCUS=security");
        assert!(args.positional.is_empty());
        assert_eq!(args.named.get("FILE"), Some(&"main.ts".to_string()));
        assert_eq!(args.named.get("FOCUS"), Some(&"security".to_string()));
    }

    #[test]
    fn test_parse_args_quoted() {
        let args = parse_args(r#"TITLE="Fix the bug" FILE=main.ts"#);
        assert_eq!(args.named.get("TITLE"), Some(&"Fix the bug".to_string()));
    }

    #[test]
    fn test_render_prompt_named() {
        let prompt = PromptDefinition {
            name: "test".to_string(),
            description: None,
            argument_hint: None,
            body: "Review $FILE focusing on $FOCUS".to_string(),
            source_path: PathBuf::new(),
            source_type: PromptSource::User,
            named_placeholders: vec!["FILE".to_string(), "FOCUS".to_string()],
            has_positional_placeholders: false,
        };
        let mut args = ParsedArgs::default();
        args.named.insert("FILE".to_string(), "main.ts".to_string());
        args.named
            .insert("FOCUS".to_string(), "security".to_string());
        let result = render_prompt(&prompt, &args);
        assert_eq!(result, "Review main.ts focusing on security");
    }

    #[test]
    fn test_render_prompt_positional() {
        let prompt = PromptDefinition {
            name: "test".to_string(),
            description: None,
            argument_hint: None,
            body: "Args: $1, $2, $3".to_string(),
            source_path: PathBuf::new(),
            source_type: PromptSource::User,
            named_placeholders: vec![],
            has_positional_placeholders: true,
        };
        let args = ParsedArgs {
            positional: vec!["foo".to_string(), "bar".to_string(), "baz".to_string()],
            named: HashMap::new(),
        };
        let result = render_prompt(&prompt, &args);
        assert_eq!(result, "Args: foo, bar, baz");
    }

    #[test]
    fn test_render_prompt_escape() {
        let prompt = PromptDefinition {
            name: "test".to_string(),
            description: None,
            argument_hint: None,
            body: "Cost is $$100".to_string(),
            source_path: PathBuf::new(),
            source_type: PromptSource::User,
            named_placeholders: vec![],
            has_positional_placeholders: false,
        };
        let result = render_prompt(&prompt, &ParsedArgs::default());
        assert_eq!(result, "Cost is $100");
    }
}
